include/revng/TupleTree/TupleTreeReference.h

```diff
@@ -123,9 +162,32 @@ class TupleTreeReference {
     }
   }
 
+  bool hasEmptyPath() const debug_function { return Path.empty(); }
+
   bool isValid() const debug_function {
-    return canGet() and not Path.empty() and get() != nullptr;
+    if (not canGet() or Path.empty())
+      return false;
+    const T *TargetPointer = getConst();
+    const T *CachedPointer = getCachedConst();
+    if (not CachedPointer)
+      return TargetPointer;
+    return TargetPointer and TargetPointer == CachedPointer;
   }
+
+  bool cacheTarget() {
+    if (isValid()) {
+      if (std::holds_alternative<const RootT *>(Root)) {
+        CachedTarget = getConst();
+      } else if (std::holds_alternative<RootT *>(Root)) {
+        CachedTarget = get();
+      } else {
+        revng_abort("Invalid root variant!");
+      }
+    }
+    return isCached();
+  }
+
+  void evictCachedTarget() { CachedTarget = static_cast<T *>(nullptr); }
```

Comment:
I see your point, but at the moment the classes that want to be `friend` with this are Model-specific, and I don't like the fact to strictly couple this with the Model.

Reply:
Are they? Isn't it `TupleTree<...>`? It should be.

Reply:
Oh right sorry. I thought you wanted to make people `friend`s of `TupleTree`, but it's totally fine to make `TupleTree` `friend` of `TupleTreeReference`. I'll do it!

Reply:
Uhm, but then if I make `TupleTree` `friend`, it'll be able to fiddle with other private fields. Like the `CachedTarget` itself, and call `private` member functions such as `canGet` and `hasNullRoot`. The latter thing doesn't worry me too much since they're both `const` methods, but if the `TupleTree` can access whatever it wants in `TupleTreeReference` it can easily invalidate the cache without using the methods, and then why are we making methods for handling cache at all.

Leaving like it is right now does not make things more dangerous. Especially if you're building TTRs from scratch outside the `TupleTree` itself there are situations where it could be beneficial to cache them as well.

Reply:
I'm more worried about a random dev (maybe downstream) using `cacheTarget` inappropriately because he has performance concerns rather than the author of `TupleTree` doing nasty things with `TupleTreeReference`.
I think we should be in full control of caching in a few, very very specific, places or we might get nasty bugs.

Maybe in the future we can relax things, if there's real need, but for now I'd err on the safe side.

But let's not overthink this, I'm not a huge fan of `friend`. If you prefer, we can keep this way.
What matters is that we don't use it.

Reply:
I hate `friend` too, but if we're thinking about guarding this against misuse from downstream developers I think it makes sense. I'll go with `friend`.

Reply:
Done